---
title: "LAB 5 Bioinformatics"
author:
- Hugo Morvan
- William Wiik
date: "`r Sys.Date()`"
output:
  pdf_document:
    latex_engine: xelatex
    fig_caption: true
    number_sections: true
  html_document:
    df_print: paged
  word_document: default
geometry: top=100pt,bottom=100pt,left=68pt,right=66pt
header-includes:
- \usepackage{float}
- \usepackage{longtable}
- \usepackage{caption}
- \usepackage{fancyhdr}
- \usepackage{titling}
- \renewcommand{\headrulewidth}{0pt}
- \renewcommand{\and}{\\}
- \pretitle{\centering\vspace{0cm}{732A51 Bioinformatics \par}\vspace{5cm}\Huge\textbf}
- \posttitle{\vspace{1cm}\large\textbf{}\par}
- \preauthor{\centering\vspace{4cm}\normalsize}
- \postauthor{\par\vspace{2cm}}
- \predate{\centering{\normalsize STIMA \\ Department of Computer and Information
  Science \\ Link√∂pings universitet \par}}
- \postdate{\par\vspace{0cm}}
- \raggedbottom
---


<!-- page number pos -->
\fancyhf{}
\fancyfoot[C]{\thepage}
\pagestyle{fancy}

<!-- no page nr on first page  -->
\pagenumbering{gobble}

<!-- Anger sidbrytning -->
\clearpage

<!-- creating the table of contents -->
\setcounter{tocdepth}{3}
\tableofcontents

<!-- new page -->
\clearpage

<!-- starting the count on 1 after the contents -->
\pagenumbering{arabic}
\setcounter{page}{1}

```{r, include = FALSE,message=FALSE}
set.seed(12345)

```


# Question 1

Go to the webpage http://snap.stanford.edu/biodata/ and choose one of the provided datasets. Download it and reproduce the statistics concerning the graph. If you obtain different values, then discuss this in your report. Visualize the graph. The next step is to try to identify some clusters (communities in the graph). You can follow the tutorial at https://psych-networks.com/r-tutorial-identify-communities-items-networks/ to achieve this. Once you have found some clusters, identify the elements in it and try to find information on this cluster. Is it related to some known biological phenomena? If you do not find anything, then document your search attempts. If it will not be possible to do this question on the whole downloaded graph, then you may take some sub-graph of it.


```{r,message=FALSE}

library(qgraph)
library(igraph)
library(ape)


con <- gzfile("ChG-Miner_miner-chem-gene.tsv.gz", "rt")
data <- read.table(con, sep = "\t", header = FALSE)
close(con)
colnames(data) <- c("Drug","Gene")

g <- graph_from_data_frame(data, directed = TRUE)
g


# Number of nodes and edges
cat("Number of nodes:", vcount(g), "\n")
cat("Number of edges:", ecount(g), "\n")


```


```{r,message=FALSE}


plot(g, vertex.size=5,vertex.label=NA, edge.width=0.01,vertex.color="coral", main="ChG-Miner_miner-chem-gene Graph")

```

```{r plot1, fig.width=16, fig.height=16, cache=TRUE}

# Function to create and plot a bipartite network from an edge matrix (We got help from Claude.ai)
plot_bipartite_network <- function(edge_matrix, title = "Bipartite Network",plot_graph = TRUE, top_n = 9) {
  # Create unique node names
  drugs <- unique(edge_matrix[,1])
  genes <- unique(edge_matrix[,2])
  
  # Create the graph
  g <- make_empty_graph(directed = FALSE)
  
  # Add all nodes
  all_nodes <- c(drugs, genes)
  g <- add_vertices(g, length(all_nodes), name = all_nodes)
  
  # Create type attribute for bipartite graph
  V(g)$type <- c(rep(TRUE, length(drugs)), rep(FALSE, length(genes)))
  V(g)$shape <- ifelse(V(g)$type, "circle", "square")

  edges <- apply(edge_matrix, 1, function(x) c(which(V(g)$name == x[1]), which(V(g)$name == x[2])))
  
  # Add all edges at once
  g <- add_edges(g, unlist(edges))


  comms <- cluster_louvain(g)
  freq_memb <- table(membership(comms))
  big_mems <- freq_memb[order(freq_memb,decreasing = T)][1:top_n]

  right_index <- as.vector(as.numeric(names(big_mems)))
  mems <- membership(comms)
  comm_col <- rainbow(top_n)[mems]
  comm_col <- rep("#FFFFFF",length(comm_col))

  topncol <- c("#FF0000", "#FFAA00", "#AAFF00", "#00FF00" ,"#00FFAA",
               "#00AAFF", "#0000FF" ,"#AA00FF", "#FF00AA")

  for(i in 1:length(right_index)) {
    comm_col[which(mems == right_index[i])] <- topncol[i]
  }
  
  # Plot the network
  if(plot_graph) {
    
    plot(g, 
         vertex.label = V(g)$name,
         vertex.shape = V(g)$shape,
         vertex.color = comm_col,
         vertex.label.cex = 0.01,  # Adjust label size
         vertex.size = 5,         # Node size
         vertex.label.color = "black",
         main = title#,
         #layout = layout_as_bipartite  # Specific layout for bipartite graphs
    )
    legend("bottomright", legend = paste("Community", right_index),
           fill = topncol)

    # Return the graph object for further analysis if needed
    return(g)
  } else {
    return(comms)
  }
}

# Example usage:
# Create a sample edge matrix (drug-gene connections)
edge_matrix <- data

# Plot the network
plot_bipartite_network(edge_matrix, "Drug-Gene Interaction Network")
```

Drug-Gene interactions are plotted. The 9 biggest communities have been visualized with colors, the rest are white. Notice that a lot of Gene-Drug interaction happen only once, resulting in a lot of isolated pairs on the outskirt of the plot. Drugs are represented by circles and Gene by Squares.

## Graph 2


```{r plot2, fig.height=16, fig.width=16, cache=TRUE}


# Function to create and plot a bipartite network from an edge matrix
plot_bipartite_network_2 <- function(edge_matrix, title = "Bipartite Network", top_n = 9) {
  # Create unique node names
  drugs <- unique(edge_matrix[,1])
  genes <- unique(edge_matrix[,2])
  
  # Create the graph
  g <- make_empty_graph(directed = FALSE)
  
  # Add all nodes
  all_nodes <- c(drugs, genes)
  g <- add_vertices(g, length(all_nodes), name = all_nodes)
  
  # Create type attribute for bipartite graph
  V(g)$type <- c(rep(TRUE, length(drugs)), rep(FALSE, length(genes)))
  
  # Add edges
  edges <- apply(edge_matrix, 1, function(x) c(which(V(g)$name == x[1]), which(V(g)$name == x[2])))
  
  # Add all edges at once
  g <- add_edges(g, unlist(edges))
  # Set node colors based on type
  #V(g)$color <- ifelse(V(g)$type, "lightblue", "lightgreen")
  
  comms <- cluster_louvain(g)
  freq_memb <- table(membership(comms))
  big_mems <- freq_memb[order(freq_memb,decreasing = T)][1:top_n]

  right_index <- as.vector(as.numeric(names(big_mems)))
  mems <- membership(comms)
  comm_col <- rainbow(top_n)[mems]
  comm_col <- rep("#FFFFFF",length(comm_col))

  topncol <- c("#FF0000", "#FFAA00", "#AAFF00", "#00FF00" ,"#00FFAA",
               "#00AAFF", "#0000FF" ,"#AA00FF", "#FF00AA")

  for(i in 1:length(right_index)) {
    comm_col[which(mems == right_index[i])] <- topncol[i]
  }
  
  
  V(g)$shape <- ifelse(V(g)$type, "circle", "square")
  # Plot the network
  
  plot(g, 
       vertex.label = V(g)$name,
       vertex.color = comm_col,
       vertex.label.cex = 0.1,  # Adjust label size
       vertex.size = 5,         # Node size
       vertex.label.color = "black",
       main = title,
       layout = layout_as_bipartite  # Specific layout for bipartite graphs
  )
  legend("bottomright", legend = paste("Community", right_index),
           fill = topncol)
  # Return the graph object for further analysis if needed
  return(g)
}

# Example usage:
# Create a sample edge matrix (drug-gene connections)
edge_matrix <- data

# Plot the network
plot_bipartite_network_2(edge_matrix, "Drug-Gene Interaction Network")
```

The Drug-Gene network can also be viewed as a bipartite graph, since Drug never connect to other drugs directly, and likewise for genes. Drugs are represented by circles and Gene by Squares.

```{r first_com, cache=TRUE}
top_n <- 9
communitys <- plot_bipartite_network(edge_matrix, plot_graph = FALSE)
freq_memb <- table(membership(communitys))
big_mems <- freq_memb[order(freq_memb,decreasing = T)][1:top_n]

right_index <- as.vector(as.numeric(names(big_mems)))

biggest_com <- communitys[[right_index[1]]]


# install.packages("BiocManager")
# BiocManager::install("clusterProfiler")
# BiocManager::install("org.Hs.eg.db")
# devtools::install_github("YuLab-SMU/clusterProfiler")

library(clusterProfiler)
library(org.Hs.eg.db)

# Perform GO enrichment analysis for the current community
go_results <- enrichGO(gene = biggest_com,
                       OrgDb = org.Hs.eg.db, keyType = "UNIPROT", ont = "BP")

top_go_terms <- go_results@result[order(go_results@result$p.adjust), ]
head(top_go_terms$Description, 10) # strongly related to neurobiology
barplot(go_results, showCategory = 10, title = "Top 10 GO terms for the biggest community")



```

the common point between these is that they are all metabolic processes.

```{r sec_com, cache=TRUE}
library(ggplot2)
sec_biggest_com <- communitys[[right_index[2]]]


# Perform GO enrichment analysis for the current community
go_results <- enrichGO(gene = sec_biggest_com,
                       OrgDb = org.Hs.eg.db, keyType = "UNIPROT", ont = "BP")

top_go_terms <- go_results@result[order(go_results@result$p.adjust), ]
head(top_go_terms$Description, 10) 
barplot(go_results, showCategory = 10, title = "Top 10 GO terms for the second biggest community") + 
  theme(axis.text.y = element_text(size = 8))


```

The common point between these seems to be that these are all signaling pathways or response mechanisms.


# Question 2


Recreate one of the three analyses that can be found on https://strimmerlab.github.io/software/genenet/index.html. Document and discuss all your steps. In the analyses there is the step where you select the edges to keep. There a particular criterion is chosen for edge inclusion. Vary this criterion and explore how the resulting clusters will differ with the changes. Take one found cluster, identify the elements in it and try to find information on this cluster. Is it related to some known biological phenomena? If you do not find anything, then document your search attempts.


## Load GeneNet package
```{r}
library("GeneNet")
library("Rgraphviz") 
```



Loading the needed dataset; E. Coli data set (9 time points for 102 genes):

```{r}
data(ecoli)
dim(ecoli)
```


## Estimation of partial correlations

Estimate matrix of partial correlation using a shrinkage estimator: A shrinkage estimator is a type of statistical method for estimating genetic effects. Shrinkage estimators involve "shrinking" or reducing the estimates towards zero, thereby reducing variability and improving precision compared to non-shrunken (ordinary) least squares estimates.

```{r}
pc = ggm.estimate.pcor(ecoli)
dim(pc)
```

Assign p-values, q-values and empirical posterior probabilities to all
5151 potential edges in the network:

```{r}
ecoli.edges = network.test.edges(pc, direct=TRUE, fdr=TRUE)
dim(ecoli.edges)
```

The table lists all edges in the order strength of partial correlations:

```{r}
ecoli.edges[1:5,]
```

## Decide which edges to include in the network

In these section are presented different methods to decide which edges to include in the final network. One method is to use various False Discovery Rate (FDR) criterion. Another method is to ahve a fix number of edges, and to only include the strongest edges.


To obtain a graph you need to select top ranking edges according to a suitable criterion.  Here are some suggestions:


1. Use local fdr cutoff 0.2, i.e. include all edges with posterior probability of at least 0.8.

```{r}
ecoli.net = extract.network(ecoli.edges)
dim(ecoli.net)
```

```{r fig.width=8, fig.height=7}
node.labels = colnames(ecoli)
gr = network.make.graph(ecoli.net, node.labels, drop.singles=TRUE)
globalAttrs = list()
globalAttrs$edge = list(color = "black", lty = "solid", lwd = 1, arrowsize=1)
globalAttrs$node = list(fillcolor = "lightblue", shape = "ellipse", fixedsize = FALSE)
 
nodeAttrs = list()
nodeAttrs$fillcolor = c('sucA' = "yellow")

edi = edge.info(gr)
edgeAttrs = list()
edgeAttrs$dir = edi$dir # set edge directions 
edgeAttrs$lty = ifelse(edi$weight < 0, "dotted", "solid") # negative correlation -> dotted
edgeAttrs$color = ifelse(edi$dir == "none", "black", "red")
edgeAttrs$label = round(edi$weight, 2) # use partial correlation as edge labels
plot(gr, attrs = globalAttrs, nodeAttrs = nodeAttrs, edgeAttrs = edgeAttrs, "fdp")

```

2. Use local fdr cutoff 0.1, i.e. i.e. include all edges with posterior probability of at least 0.9.


```{r}
ecoli.net = extract.network(ecoli.edges, cutoff.ggm=0.9, cutoff.dir=0.9)
dim(ecoli.net)
```

```{r fig.width=8, fig.height=7}
gr = network.make.graph(ecoli.net, node.labels, drop.singles=TRUE)
globalAttrs = list()
globalAttrs$edge = list(color = "black", lty = "solid", lwd = 1, arrowsize=1)
globalAttrs$node = list(fillcolor = "lightblue", shape = "ellipse", fixedsize = FALSE)
 
nodeAttrs = list()
nodeAttrs$fillcolor = c('sucA' = "yellow")

nodeAttrs$fillcolor = c('cspA' = "green",'hupB' = "green",'yfiA' = "green")

edi = edge.info(gr)
edgeAttrs = list()
edgeAttrs$dir = edi$dir # set edge directions 
edgeAttrs$lty = ifelse(edi$weight < 0, "dotted", "solid") # negative correlation -> dotted
edgeAttrs$color = ifelse(edi$dir == "none", "black", "red")
edgeAttrs$label = round(edi$weight, 2) # use partial correlation as edge labels
plot(gr, attrs = globalAttrs, nodeAttrs = nodeAttrs, edgeAttrs = edgeAttrs, "fdp")

```

3. Include a fixed number of edges, say the 70 strongest edges

```{r}
ecoli.net = extract.network(ecoli.edges, method.ggm="number", cutoff.ggm=70)
dim(ecoli.net)
```

```{r fig.width=8, fig.height=7}
gr = network.make.graph(ecoli.net, node.labels, drop.singles=TRUE)
globalAttrs = list()
globalAttrs$edge = list(color = "black", lty = "solid", lwd = 1, arrowsize=1)
globalAttrs$node = list(fillcolor = "lightblue", shape = "ellipse", fixedsize = FALSE)
 
nodeAttrs = list()
nodeAttrs$fillcolor = c('sucA' = "yellow")

nodeAttrs$fillcolor = c('cspA' = "green",'hupB' = "green",'yfiA' = "green")

edi = edge.info(gr)
edgeAttrs = list()
edgeAttrs$dir = edi$dir # set edge directions 
edgeAttrs$lty = ifelse(edi$weight < 0, "dotted", "solid") # negative correlation -> dotted
edgeAttrs$color = ifelse(edi$dir == "none", "black", "red")
edgeAttrs$label = round(edi$weight, 2) # use partial correlation as edge labels
plot(gr, attrs = globalAttrs, nodeAttrs = nodeAttrs, edgeAttrs = edgeAttrs, "fdp")

```

Plot network

For plotting we use the graph and Rgraphviz packages from Bioconductor.

```{r warning=FALSE}
library("Rgraphviz") 
```

Create graph object from the list of edges:

```{r}
node.labels = colnames(ecoli)
gr = network.make.graph(ecoli.net, node.labels, drop.singles=TRUE)
table(  edge.info(gr)$dir )
sort( node.degree(gr), decreasing=TRUE)
```


Set node and edge attributes for more beautiful graph plotting:

```{r}
globalAttrs = list()
globalAttrs$edge = list(color = "black", lty = "solid", lwd = 1, arrowsize=1)
globalAttrs$node = list(fillcolor = "lightblue", shape = "ellipse", fixedsize = FALSE)
 
nodeAttrs = list()
nodeAttrs$fillcolor = c('sucA' = "yellow")

nodeAttrs$fillcolor = c('cspA' = "green",'hupB' = "green",'yfiA' = "green")

edi = edge.info(gr)
edgeAttrs = list()
edgeAttrs$dir = edi$dir # set edge directions 
edgeAttrs$lty = ifelse(edi$weight < 0, "dotted", "solid") # negative correlation -> dotted
edgeAttrs$color = ifelse(edi$dir == "none", "black", "red")
edgeAttrs$label = round(edi$weight, 2) # use partial correlation as edge labels
```


```{r fig.width=8, fig.height=7}
plot(gr, attrs = globalAttrs, nodeAttrs = nodeAttrs, edgeAttrs = edgeAttrs, "fdp")

```


We observe that low values for the FDR cutoff value produces less significant edges in the network. We can demonstrate the opposite by having a higher FDR cutoff, which results in many more significant edges, and thus a denser graph.


```{r}
ecoli.net = extract.network(ecoli.edges, cutoff.ggm=0.5, cutoff.dir=0.5)
dim(ecoli.net)
```


```{r fig.width=8, fig.height=7}
gr = network.make.graph(ecoli.net, node.labels, drop.singles=TRUE)
globalAttrs = list()
globalAttrs$edge = list(color = "black", lty = "solid", lwd = 1, arrowsize=1)
globalAttrs$node = list(fillcolor = "lightblue", shape = "ellipse", fixedsize = FALSE)
 
nodeAttrs = list()
nodeAttrs$fillcolor = c('sucA' = "yellow")
nodeAttrs$fillcolor = c('cspA' = "green",'hupB' = "green",'yfiA' = "green")
#cspA - hupB - yfiA
edi = edge.info(gr)
edgeAttrs = list()
edgeAttrs$dir = edi$dir # set edge directions 
edgeAttrs$lty = ifelse(edi$weight < 0, "dotted", "solid") # negative correlation -> dotted
edgeAttrs$color = ifelse(edi$dir == "none", "black", "red")
edgeAttrs$label = round(edi$weight, 2) # use partial correlation as edge labels
plot(gr, attrs = globalAttrs, nodeAttrs = nodeAttrs, edgeAttrs = edgeAttrs, "fdp")
```

## Cluster analysis:

Finding a common factor between these: 

cspA - hupB - yfiA triangle/cluster in E. Coli data. 

yfiA: https://ecocyc.org/gene?orgid=ECOLI&id=EG11151

hupB: https://ecocyc.org/gene?orgid=ECOLI&id=EG10467

cspA: https://ecocyc.org/gene?orgid=ECOLI&id=EG10166 

All three genes appear to be involved in complex regulatory processes related to gene expression, protein interactions, or cellular stress responses. They also seems to have a role in stress response, particularly cold shock and is related to hibernation factors. It is hard to have a better understanding of these proteins' functions without a deeper biologic understanding, but we think that this cluster plays a role in protecting E. Coli in cold environment. 
