---
title: "LAB 2 Bioinformatics"
author:
- Hugo Morvan
- William Wiik
date: "`r Sys.Date()`"
output:
  pdf_document:
    latex_engine: xelatex
    fig_caption: true
    number_sections: true
  html_document:
    df_print: paged
  word_document: default
geometry: top=100pt,bottom=100pt,left=68pt,right=66pt
header-includes:
- \usepackage{float}
- \usepackage{longtable}
- \usepackage{caption}
- \usepackage{fancyhdr}
- \usepackage{titling}
- \renewcommand{\headrulewidth}{0pt}
- \renewcommand{\and}{\\}
- \pretitle{\centering\vspace{0cm}{732A51 Bioinformatics \par}\vspace{5cm}\Huge\textbf}
- \posttitle{\vspace{1cm}\large\textbf{}\par}
- \preauthor{\centering\vspace{4cm}\normalsize}
- \postauthor{\par\vspace{2cm}}
- \predate{\centering{\normalsize STIMA \\ Department of Computer and Information
  Science \\ Linköpings universitet \par}}
- \postdate{\par\vspace{0cm}}
- \raggedbottom
---


<!-- page number pos -->
\fancyhf{}
\fancyfoot[C]{\thepage}
\pagestyle{fancy}

<!-- no page nr on first page  -->
\pagenumbering{gobble}

<!-- Anger sidbrytning -->
\clearpage

<!-- creating the table of contents -->
\setcounter{tocdepth}{3}
\tableofcontents

<!-- new page -->
\clearpage

<!-- starting the count on 1 after the contents -->
\pagenumbering{arabic}
\setcounter{page}{1}

```{r, include = FALSE,message=FALSE}
set.seed(12345)
library(dplyr)
library(ape)
library(seqinr)
library(msa)
library(Biostrings)
```


# Question 1

In this exercise you will perform statistical analysis of three nucleotide data sets. First download the sequences from GenBank and save them in a fasta le. For this use the provided R script, `732A51_BioinformaticsHT2023_Lab02_GenBankGetCode.R`. This is a dataset of the RAG1 gene sequences from 33 lizard species. You are encouraged to read in detail the references in the script as they indicate many useful tools. Explore the dataset using the tools provided by the `ape` and `seqinr` packages. Take note of the lengths of all the sequences and the base composition.


```{r 1-0}

## Gene bank accession numbers taken from http://www.jcsantosresearch.org/Class_2014_Spring_Comparative/pdf/week_2/Jan_13_15_2015_GenBank_part_2.pdf
lizards_accession_numbers <- c("JF806202", "HM161150", "FJ356743", "JF806205", 
                               "JQ073190", "GU457971", "FJ356741", "JF806207",
                               "JF806210", "AY662592", "AY662591", "FJ356748",       
                               "JN112660", "AY662594", "JN112661", "HQ876437", 
                               "HQ876434", "AY662590", "FJ356740", "JF806214", 
                               "JQ073188", "FJ356749", "JQ073189", "JF806216", 
                               "AY662598", "JN112653", "JF806204", "FJ356747", 
                               "FJ356744", "HQ876440", "JN112651", "JF806215",
                               "JF806209") 
lizards_sequences<-ape::read.GenBank(lizards_accession_numbers)
print(lizards_sequences)
ape::write.dna(lizards_sequences, file ="lizard_seqs.fasta", format = "fasta",
               append =FALSE, nbcol = 6, colsep = " ", colw = 10)


```


## Question 1.1

Simulate an artificial DNA sequence dataset. It should contain 33 sequence. The lengths of the sequences should be the same as in the lizard dataset, i.e. for each real sequence simulate an artffcial one. The simulation rule is as follows, each nucleotide is to be independently and randomly drawn from the distribution given by the base composition (frequencies) in the true lizard sequences. Save your dataset in a fasta format file. Remember to give unique names to your sequences. Report on the base composition in your simulated data.

**Answer:**

```{r 1-1}

true_base_composition <- c(0.312, 0.205, 0.231, 0.252) # a c g t

# Length of each real sequence
sequence_lengths <- sapply(lizards_sequences, length)

# Simulating DNA
sim_sequences <- lapply(sequence_lengths, function(length) {
   rDNAbin(n = length, base.freq = true_base_composition)[[1]]
})

names(sim_sequences) <- paste0("sim_",names(lizards_sequences))

sim_sequences <- structure(sim_sequences, class = "DNAbin")

ape::write.dna(sim_sequences, file = "sim_lizard_seqs.fasta", 
               format = "fasta", append = FALSE, nbcol = 6, colsep = " ", colw = 10)

sim_lizard_seqs <- read.FASTA("sim_lizard_seqs.fasta")
print(sim_lizard_seqs)

```



a: 31.2%

c: 20.3%

g: 23.0%

t: 25.5%

We obtain almost the same base composition as the true base composition for lizard sequences, which is to be expected.


## Question 1.2 *

```{r 1-2}

```


\newpage


# Question 2

## Question 2.1

Report some basic statistics on each sequence dataset: individual base composition, $GC$ content, $CG$, $AT$ content. Also translate your sequences into protein sequences (see Lab 1) and report on the amino acid composition. In your simulated sequences, how many times did you observe a stop codon inside your sequence? Does this occur in your true sequences? Comment.


```{r 2-1}

true_sequences <- read.FASTA("lizard_seqs.fasta")
sim_sequences <- read.FASTA("sim_lizard_seqs.fasta")

calc_seq_stats <- function(all_seq, index_seq) {
  

 seq_DNAbin  <- structure(all_seq[[index_seq]], class = "DNAbin")
 seq <- as.character(seq_DNAbin) # from raw to character
 
 
 # How many g and c that are in the sequence
 GC_content <- ((sum(seq == "g") + sum(seq == "c")) / length(seq)) 
 
 # Base composition
 base <- base.freq(seq_DNAbin) %>% round(3)
 
 # CpG (How many c is followed by a g)
 # seq[-length(seq)] removes the last element
 # seq[-1] removes the first element
 CpG <- sum(seq[-length(seq)] == "c" & seq[-1] == "g") / length(seq)
 
 
 # ApT (How many a is followed by a t)
 ApT <- sum(seq[-length(seq)] == "a" & seq[-1] == "t") / length(seq)
 
 
 
 print(paste0("Base composition for ", names(all_seq)[index_seq],
              ": a=",base["a"],", c=",base["c"],", g=",base["g"],", t=",base["t"]))
 
 print(paste0("GC content for ", names(all_seq)[index_seq],": ",round(GC_content,3)))
 
 print(paste0("CpG content for ", names(all_seq)[index_seq],": ",round(CpG,3)))
 
 print(paste0("ApT content for ", names(all_seq)[index_seq],": ",round(ApT,3)))
 
}

```

```{r 2-1-2}
# True lizard sequence 1 (JF806202)
calc_seq_stats(all_seq = true_sequences, index_seq = 1)
```

```{r 2-1-3}
# Simulate lizard sequence 1 (sim_JF806202)
calc_seq_stats(all_seq = sim_sequences, index_seq = 1)
```

```{r 2-1-4}

# "Cleanning" the sequences because they were some issues on my personal computer
dna_seqs <- readDNAStringSet("lizard_seqs.fasta")
cleaned_seqs <- DNAStringSet(gsub("[^ATCG]", "", dna_seqs))


dna_seqs_sim <- readDNAStringSet("sim_lizard_seqs.fasta")
cleaned_seqs_sim <- DNAStringSet(gsub("[^ATCG]", "", dna_seqs_sim))

# File from Ying Luo
codon_and_ORF <- read.csv("codon_and_ORF.csv")

vec <- c()
for(i in 1:33){
  start_codon<- codon_and_ORF$codon_start[i]
  seq = cleaned_seqs[[i]]
  seq_vec = unlist(strsplit(tolower(as.character(seq)),''))
  translation = seqinr::translate(seq_vec, start_codon-1)
  vec <- c(vec, translation)
}

char_table <- prop.table(table(vec))
round(char_table[order(table(vec))],2)

vec_sim <- c()
for(i in 1:33){
  start_codon<- codon_and_ORF$codon_start[i]
  seq = cleaned_seqs_sim[[i]]
  seq_vec = unlist(strsplit(tolower(as.character(seq)),''))
  translation = seqinr::translate(seq_vec)
  vec_sim <- c(vec_sim, translation)
}

char_table <- prop.table(table(vec_sim))
round(char_table[order(table(vec_sim))],2)


```

For the true sequences, we get almost no stopping codons (*), while for the simulated sequences, we get a proportion of stopping codons proportional to the number of ways to obtain a stopping codon from the standard codon table. This is to be expected since the simulated sequences generate proteins at random. For the true sequences, we expect to have much less stopping codons since they should only appear at the end of the protein.

## Question 2.2*

```{r 2-2}

```


## Question 2.3

Align your sequences using software of your choice (a starter for R: https://stackoverflow.com/questions/4497747/how-to-perform-basic-multiple-sequence-alignments-in-r, you can also look what `Biopython`, `BioPerl` offer, use the `Clustal` family of programs or something else of your choice). Choose a distance measure between sequences, calculate for each alignment the distances between all pairs of sequences. Then, plot heatmaps visualizing the distances. Comment on what you can observe.

```{r 2-3-0, include=FALSE}
if (!requireNamespace("msa", quietly = TRUE)) {
  install.packages("BiocManager")
  BiocManager::install("msa")
}
```


```{r 2-3-1}

clust_lizard_seqs <- msaClustalW(cleaned_seqs, type = "dna")
clust_sim_lizard_seqs <- msaClustalW(cleaned_seqs_sim, type = "dna")

# Convert MSA object to formats used in other sequence
# type = "seqinr::alignment" because we want to use dist.alignment() later
converted_lizard_seqs <- msaConvert(clust_lizard_seqs, type = "seqinr::alignment")
converted_sim_lizard_seqs <- msaConvert(clust_sim_lizard_seqs, type = "seqinr::alignment")


# Calculate pairwise distances using dist.alignment() from seqinr package
dist_matrix_lizard_seqs <- as.matrix(dist.alignment(converted_lizard_seqs), matrix = "identity")
dist_matrix_sim_lizard_seqs <- as.matrix(dist.alignment(converted_sim_lizard_seqs), matrix = "identity")

```


```{r 2-3-2}
heatmap(dist_matrix_lizard_seqs, main = "Distance between sequences (True lizard seq.)")
```



```{r 2-3-3}
heatmap(dist_matrix_sim_lizard_seqs, main = "Distance between sequences (Simulated seq.)")
```

**Answer**

We observer some similarities between genes in the true data (heat on the heatmap), while for the simulated data, there is practically no similarity. This is expected because the simulated data comes from a iid process, while the true data comes from genetic evolution.

# Question 3


## Question 3.1

Construct (using algorithm and software of your choice) phylogenetic trees from the three multiple alignments (or distance matrices) done in Question 2.3. You might want to look at the functions offered by `ape`, `phangorn` (https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf) or go for some completely different software. Plot the inferred trees. Are the two based on the simulated data similar to expected?
Perform a phylogenetic bootstrap analysis and report the bootstrap support for the individual clades, you can look at `ape::boot.phylo()`.

```{r 3-1}
# Neighbor-Joining Tree Estimation :
phyl1 <- nj(dist_matrix_lizard_seqs)
plot(phyl1)
#Other options to create the tree : fastme.bal bionj, mvr(X, V), nj 
```

```{r}
# Neighbor-Joining Tree Estimation :
phyl2 <-nj(dist_matrix_sim_lizard_seqs)
plot(phyl2)
```
```{r}
#Boostrapping
## from ?boot.phylo:
f <- function(x) nj(x)
tw <- f(dist_matrix_lizard_seqs)
## bootstrap with 1000 replications:
(bp <- boot.phylo(tw, dist_matrix_lizard_seqs, f, B=100, quiet=TRUE))
## more readable but the tree is really unrooted:
plot(tw)
drawSupportOnEdges(bp)


```



```{r}
#Boostrapping
## from ?boot.phylo:
f <- function(x) nj(x)
tw <- f(dist_matrix_sim_lizard_seqs)
## bootstrap with 100 replications:
(bp <- boot.phylo(tw, dist_matrix_sim_lizard_seqs, f, B=100, quiet=TRUE))
plot(tw)
drawSupportOnEdges(bp)
```

Interpretation Scale:

70-95%: Moderately supported branch
95-100%: Highly supported branch
<70%: Weak support, less confidence in that branching pattern

For both true and simulated trees, none of the branches have significant suppport (>70%, i.e. support>70)
Tree building functions tested : fastme.bal, bionj, nj., all yielded the same results.


## Question 3.2*

Compare your inferred trees and also your simulated one. Apart from visualizing the trees one may calculate various indices related to them and distances between the trees. Explore what indices and metrics the `ape`, `distory`, `phangorn`, `phyloTop`, `TotalCopheneticIndex` or `treespace` R packages offer, choose some and report the results in a meaningful way. You might have to save your tree to drive and then read in it using e.g. ape’s tree reading functionality.

```{r 3-2}
ape::comparePhylo(phyl1, phyl2)
```


```{r 3-3}
#Metrics : 

branching.times(phyl1)

branching.times(phyl2)
```


```{r 3-4}
# #compar.cheverud         Cheverud's Comparative Method
# #compar.gee              Comparative Analysis with GEEs
# #compar.lynch            Lynch's Comparative Method
# #compar.ou               Ornstein-Uhlenbeck Model for Continuous Characters
# 
# a = compute.brlen(phyl1)           #Branch Lengths Computation
# plot(a)
# b = compute.brtime(phyl1)          #Compute and Set Branching Times
# plot(b)    
# 
# a = compute.brlen(phyl2)           #Branch Lengths Computation
# plot(a)
# b = compute.brtime(phyl2)          #Compute and Set Branching Times
# plot(b)    
# 
# heatmap(cophenetic.phylo(phyl1))
# heatmap(cophenetic.phylo(phyl2))

# corBlomberg             #Blomberg et al.'s Correlation Structure
# corBrownian             #Brownian Correlation Structure
# corClasses              #Phylogenetic Correlation Structures
# corGrafen               #Grafen's (1989) Correlation Structure
# corMartins              #Martins's (1997) Correlation Structure
# corPagel                #Pagel's "lambda" Correlation Structure
# corphylo                #Correlations among Multiple Traits with Phylogenetic Signal
# correlogram.formula     #Phylogenetic Correlogram
# 
# dist.topo(phyl1, phyl2) #needs similar leaf names
# 
# gammaStat               #Gamma-Statistic of Pybus and Harvey, requires ultrametric trees
# 
# root                    #Roots Phylogenetic Trees

summary.phylo(phyl1)
summary.phylo(phyl2)
```

**Answer:**

*True data phylogenetic tree Branch lengths stats:*

mean: 0.05191795 , variance: 0.001511728 

Min.: 0.001377696

1st Qu.: 0.014389981 

Median: 0.051113051

3rd Qu.:  0.075027992

Max.: 0.138188147


*Simulated data phylogenetic tree Branch lengths stats:*

mean: 0.2178854 , variance: 0.04231989 

Min.: 0.0005453872 

1st Qu.: 0.0027112159 

Median: 0.3607140035 

3rd Qu. : 0.4162849315  

Max.: 0.4275080112 

The simulated data phylogeneteic tree branch length has on average much longer branch lengths (0.05 vs 0.22), but with also much more variance (0.0015 vs 0.0423). These results correlate with the observation from the heatmaps which shows that in the simulated data, no clade is close to any other, which makes sense since the data was randomly generated and should be i.i.d. 
