---
title: "LAB 3 Bioinformatics"
author:
- Hugo Morvan
- William Wiik
date: "`r Sys.Date()`"
output:
  pdf_document:
    latex_engine: xelatex
    fig_caption: true
    number_sections: true
  html_document:
    df_print: paged
  word_document: default
geometry: top=100pt,bottom=100pt,left=68pt,right=66pt
header-includes:
- \usepackage{float}
- \usepackage{longtable}
- \usepackage{caption}
- \usepackage{fancyhdr}
- \usepackage{titling}
- \renewcommand{\headrulewidth}{0pt}
- \renewcommand{\and}{\\}
- \pretitle{\centering\vspace{0cm}{732A51 Bioinformatics \par}\vspace{5cm}\Huge\textbf}
- \posttitle{\vspace{1cm}\large\textbf{}\par}
- \preauthor{\centering\vspace{4cm}\normalsize}
- \postauthor{\par\vspace{2cm}}
- \predate{\centering{\normalsize STIMA \\ Department of Computer and Information
  Science \\ Link√∂pings universitet \par}}
- \postdate{\par\vspace{0cm}}
- \raggedbottom
---


<!-- page number pos -->
\fancyhf{}
\fancyfoot[C]{\thepage}
\pagestyle{fancy}

<!-- no page nr on first page  -->
\pagenumbering{gobble}

<!-- Anger sidbrytning -->
\clearpage

<!-- creating the table of contents -->
\setcounter{tocdepth}{3}
\tableofcontents

<!-- new page -->
\clearpage

<!-- starting the count on 1 after the contents -->
\pagenumbering{arabic}
\setcounter{page}{1}

```{r, include = FALSE,message=FALSE}
set.seed(12345)
library(dplyr)
library(ape)
```


# Question 1

Using the script http://ape-package.ird.fr/APER/APER2/SylviaWarblers.R obtain the *Sylvia* warblers phylogeny (the script saves in in the file sylvia_nj_k80.tre). The geographical range data can be found in http://ape-package.ird.fr/APER/APER2/sylvia_data.txt and in the script is referenced as `DF$geo.range`. Notice that one tip is removed due to missing data

```{r, sylvia, , eval=FALSE, include=FALSE, cache=TRUE}
library(ape)

### 
### Chapter 3
###

x <- paste("AJ5345", 26:49, sep = "")
x <- c("Z73494", x)
sylvia.seq <- read.GenBank(x)

sylvia.clus <- clustal(sylvia.seq)
library(phyloch)
sylvia.maff <- mafft(sylvia.seq, path = "/usr/bin/mafft")
identical(sylvia.clus[x, ], sylvia.maff[x, ])

taxa.sylvia <- attr(sylvia.seq, "species")
names(taxa.sylvia) <- names(sylvia.seq)
rm(sylvia.seq)
taxa.sylvia[1] <- "Sylvia_atricapilla"
taxa.sylvia[24] <- "Sylvia_abyssinica"

sylvia.eco <- read.table("sylvia_data.txt")
str(sylvia.eco)
rownames(sylvia.eco)
save(sylvia.clus, taxa.sylvia, sylvia.eco,
     file = "sylvia.RData")

###
### Chapter 5
###
sylvia.seq.ali<-sylvia.clus
syl.K80 <- dist.dna(sylvia.seq.ali, pairwise.deletion = TRUE)
syl.F84 <- dist.dna(sylvia.seq.ali, model = "F84", p = TRUE)
syl.TN93 <- dist.dna(sylvia.seq.ali, model = "TN93", p = TRUE)
syl.GG95 <- dist.dna(sylvia.seq.ali, model = "GG95", p = TRUE)

round(cor(cbind(syl.K80, syl.F84, syl.TN93, syl.GG95)), 3)

syl.JC69 <- dist.dna(sylvia.seq.ali, model = "JC69", p = TRUE)
syl.raw <- dist.dna(sylvia.seq.ali, model = "raw", p = TRUE)
layout(matrix(1:2, 1))
plot(syl.JC69, syl.raw)
abline(b = 1, a = 0) # draw x = y line
plot(syl.K80, syl.JC69)
abline(b = 1, a = 0)

layout(matrix(1:3, 1))
for (i in 1:3) {
    s <- logical(3); s[i] <- TRUE
    x <- sylvia.seq.ali[, s]
    d <- dist.dna(x, p = TRUE)
    ts <- dist.dna(x, "Ts", p = TRUE)
    tv <- dist.dna(x, "Tv", p = TRUE)
    plot(ts, d, xlab = "Number of Ts or Tv", col = "blue",
         ylab = "K80 distance", xlim = range(c(ts, tv)),
         main = paste("Position", i))
    points(tv, d, col = "red")
}

y <- numeric()
for (i in 1:3) {
    s <- logical(3); s[i] <- TRUE
    y <- c(y, dist.dna(sylvia.seq.ali[, s], p = TRUE))
}
g <- gl(3, length(y) / 3)
library(lattice)
histogram(~ y | g, breaks = 20)

nj.sylvia.K80 <- nj(syl.K80)
nj.sylvia.GG95 <- nj(syl.GG95)
dist.topo(nj.sylvia.K80, nj.sylvia.GG95)

grep("Chamaea", taxa.sylvia, value = TRUE)
f <- function(xx) root(nj(dist.dna(xx, p=TRUE)), "AJ534526")
tr <- f(sylvia.seq.ali)
## same than: tr <- root(nj.sylvia.K80, "AJ534526")
nj.boot.sylvia <- boot.phylo(tr, sylvia.seq.ali, f, 200,
                             rooted = TRUE)
nj.boot.codon <- boot.phylo(tr, sylvia.seq.ali, f, 200, 3,
                            rooted = TRUE)
nj.est <- tr
nj.est$tip.label <- taxa.sylvia[tr$tip.label]
plot(nj.est, no.margin = TRUE)
nodelabels(round(nj.boot.sylvia / 200, 2), bg = "white")
add.scale.bar(length = 0.01)
write.tree(nj.est, "sylvia_nj_k80.tre")

write.dna(sylvia.seq.ali, "sylvia.txt")
phyml.sylvia <- phymltest("sylvia.txt", execname = "/home/h/Documents/phylml/phyml/src/phyml")
summary(phyml.sylvia)

plot(phyml.sylvia, col = "black")
TR <- read.tree("sylvia.txt_phyml_tree.txt")
mltr.sylvia <- TR[[28]]
mltr.sylvia$tip.label <- taxa.sylvia[mltr.sylvia$tip.label]
mltr.sylvia <- root(mltr.sylvia, "Chamaea_fasciata")
plot(mltr.sylvia, no.margin = TRUE)
add.scale.bar(length = 0.01)

tr.ml <- drop.tip(mltr.sylvia, "Chamaea_fasciata")
res <- vector("list", 9)
for (L in -4:4)
    res[[L + 5]] <- chronopl(tr.ml, 10^L, 12, 16, CV = TRUE)
Lambda <- 10^(-4:4)
CV <- sapply(res, function(x) sum(attr(x, "D2")))
plot(Lambda, CV / 1e5, log = "x")

sylvia.chrono <- res[[2]]
rts <- attr(sylvia.chrono, "rates")
summary(rts)

par(mar = c(2, 0, 0, 0))
plot(sylvia.chrono, edge.width = 100*rts, label.offset = .15)
axisPhylo()
write.tree(sylvia.chrono, "sylvia.chrono.tre")


```




## Question 1.1

Explain all the steps in the script required to obtain the phylogeny and trait data.


**Answer:**

Load the libraries:
```{r a, cache=TRUE}
library(ape)
library(phyloch)
```

Read the nucleotid sequence:
```{r b, cache=TRUE}
x <- paste("AJ5345", 26:49, sep = "")
x <- c("Z73494", x)
sylvia.seq <- read.GenBank(x)
```

Allign the sequences
```{r}
sylvia.clus <- clustal(sylvia.seq)
library(phyloch)
sylvia.maff <- mafft(sylvia.seq, path = "/usr/bin/mafft")
identical(sylvia.clus[x, ], sylvia.maff[x, ]) #check that the result are equivalent

```


Obtain species names and get rid of the rest:
```{r c, cache=TRUE}
taxa.sylvia <- attr(sylvia.seq, "species")
names(taxa.sylvia) <- names(sylvia.seq)
rm(sylvia.seq)
taxa.sylvia[1] <- "Sylvia_atricapilla"
taxa.sylvia[24] <- "Sylvia_abyssinica"
```
Read data from text file, then save the data:
```{r d, cache=TRUE}
sylvia.eco <- read.table("sylvia_data.txt")
str(sylvia.eco)
rownames(sylvia.eco)
save(sylvia.clus, taxa.sylvia, sylvia.eco,
     file = "sylvia.RData")
```

Load the DNA sequences, and calculate pairwise distance matrices from the DNA sequences using various DNA evolutionary models (K80, F84, TN93, GG95):
```{r e, cache=TRUE}
sylvia.seq.ali<-sylvia.maff #or sylvia.clus or sylvia.eco
syl.K80 <- dist.dna(sylvia.seq.ali, pairwise.deletion = TRUE)
syl.F84 <- dist.dna(sylvia.seq.ali, model = "F84", p = TRUE)
syl.TN93 <- dist.dna(sylvia.seq.ali, model = "TN93", p = TRUE)
syl.GG95 <- dist.dna(sylvia.seq.ali, model = "GG95", p = TRUE)
```
Print correlation between the different model estimations:
```{r f, cache=TRUE}
round(cor(cbind(syl.K80, syl.F84, syl.TN93, syl.GG95)), 3)
```

Bootstrap to Estimate significance of tree generating by Neighbor-Joining method. 

```{r k, cache=TRUE}

f <- function(xx) root(nj(dist.dna(xx, p=TRUE)), "AJ534526")
tr <- f(sylvia.seq.ali)
## same than: tr <- root(nj.sylvia.K80, "AJ534526")

# Estimate tree significance using bootstrap
nj.boot.sylvia <- boot.phylo(tr, sylvia.seq.ali, f, 200,
                             rooted = TRUE)

nj.est <- tr #Neighbor Joining estimate
nj.est$tip.label <- taxa.sylvia[tr$tip.label]
plot(nj.est, no.margin = TRUE)
nodelabels(round(nj.boot.sylvia / 200, 2), bg = "white")
add.scale.bar(length = 0.01)

#Save the tree
write.tree(nj.est, "sylvia_nj_k80.tre")
```


```{r preprocess, include=FALSE}
# 1. Load tree
tr <- read.tree("sylvia_nj_k80.tre")

# 2. Remove species
tr <- drop.tip(tr, "Chamaea_fasciata")
tr <- drop.tip(tr, "Devioeca_papuana")

# 3. Unify names
tr$tip.label <- gsub("Curruca_", "Sylvia_", tr$tip.label)

# 4. Fix spelling errors
tr$tip.label[tr$tip.label == "Sylvia_ruppeli"] <- "Sylvia_rueppelli"

# 5. Extract matched ecological data
DF <- sylvia.eco[tr$tip.label, ]

```


## Question 1.2 *

Analyze the discrete `(type=discrete)`geographical range variable (`DF$geo.range`) using `ape::ace`. Consider different models (parameter model). Report on the results and interpret the estimated rates and their standard errors.

```{r 1-2}
x <- factor(DF$geo.range)

rooted <- multi2di(tr)
plot(rooted)
#Equal Rate (default)
ans <- ace(x, rooted, type = "discrete")
print(ans)

#Symmetrical Model
ans <- ace(x, rooted, type = "discrete", model = "SYM")
print(ans)

#All rates Different Model
ans <- ace(x, rooted, type = "discrete", model = "ARD")
print(ans)
```


We are trying to estimate the trait of the ancestor species at the root. 

Using the model "Equal Rate" , we obtain Log-likelihood: -20.89321, The estimated rate is 5.1813 +/- 1.7578 . The scaled log-likelihood at the root is:

temp: 0.0147, temptrop : 0.9401, trop : 0.0451, 

meaning that the model estimates the most likely trait to be temptrop for the root ancestor.



Using the model "Symmetric", we obtain Log-likelihood: -19.52549. The estimated rates are:

- temp <-> temptrop rate : 3.4054 +/- 1.8762
- temp <-> trop rate : 0 (NaN std err)
- temptrop <-> trop rate : 9.3672 +/- 4.1510

The scaled log-likelihood at the root is: 

temp: 0.00174, temptrop: 0.8397, trop: 0.1585

meaning that the model estimates the most likely trait to be temptrop for the root ancestor.

- temp -> temptrop: 2.7162 +/- 2.0550
- temp -> trop: 0 +/- 15.5967
- temptrop -> temp: 0 +/- 31.8116
- temptrop -> trop: 0 +/- NaN
- trop -> temp: 0 +/- 26.6413
- trop -> temptrop:  6.5245 +/- 3.9666



Using the model "All rates different between states", we obtain Log-likelihood: -20.07269. The estimated rates are:

The scaled log-likelihood at the root is: 

temp: 0, temptrop: 1, trop: 0, 

meaning that the model is sure that the trait at the root was temptrop.


Based on the likelihood, the best model is the Symmetric model, however, all three models estimate the ancestral character (trait) to be "temptrop."

\newpage


# Question 2

Install the `ade4` package. Included with it you will find the carnivores dataset, `data(carni70)`

## Question 2.1

Explore the data set and report what can be found in it. Provide some plots.


```{r 2-1}
library(ade4)
library(ggplot2)
library(cowplot)

data(carni70)

tab_df <- as.data.frame(carni70$tab)
summary(tab_df)


p1 <- ggplot(tab_df, aes(x = size)) + 
  geom_histogram(binwidth = 5, fill = "skyblue", color = "black", alpha = 0.7) +
  labs(title = "Histogram of Size", x = "Size", y = "Frequency") + theme_bw()

p2 <- ggplot(tab_df, aes(x = range)) + 
  geom_histogram(binwidth = 1, fill = "salmon", color = "black", alpha = 0.7) +
  labs(title = "Histogram of Range", x = "Range", y = "Frequency") + theme_bw()

p3 <- ggplot(tab_df, aes(x = size, y = range)) + 
  geom_point(color = "blue") + 
  labs(title = "Scatter Plot of Size vs Range", x = "Size", y = "Range") + theme_bw()


cowplot::plot_grid(p1, p2, p3, ncol = 2)  




```

```{r, echo=F}

print(paste("Total number of carnivores:",nrow(tab_df)))

print(paste("The carnivore with biggest size:",rownames(tab_df)[which.max(tab_df$size)]))
print(paste("The carnivore with smallest size:",rownames(tab_df)[which.min(tab_df$size)]))

print(paste("The carnivore with biggest range:",rownames(tab_df)[which.max(tab_df$range)]))
print(paste("The carnivore with smallest range:",rownames(tab_df)[which.min(tab_df$range)]))
```
There are 70 carnivores, with a median size of 3.2 and a median range of 6.1. Two clear outliers in size are Ursus arctos (Brown bear) and Tremarctos ornatus (Spectacled bear).



## Question 2.2*
